lib/Doctrine/MongoDB/EagerCursor.php

```diff
@@ -237,6 +296,26 @@ public function setReadPreference($readPreference, array $tags = null)
     }
 
     /**
+     * {@inheritDoc}
+     */
+    public function hasNext()
+    {
+        $this->initialize();
+
+        $wasIterating = $this->iterating;
+        $hasNext = $this->getNext() !== null;
+
+        // Reset the internal cursor if we weren't iterating
+        if ($wasIterating) {
+            prev($this->data);
+        } else {
+            $this->iterating = false;
```

Comment:
```iterating``` has to be reset to false, otherwise calling ```hasNext()``` followed by ```getNext()``` will return the second result instead of the first result as would be expected.

If the cursor is marked as "iterating", ```getNext()``` will cause the ```next``` function which will advance the array pointer to the second result and return this. The first result however has never been fetched and would be skipped.